### DISCUSSION TITLE
Weird problem with multiple interpreter services...

### DISCUSSION THREAD

**Discussion by vck3000** [statelyai/xstate#1313] [General] [answered]
Sorry if this isn't the appropriate place to ask this, but I can't seem to understand why this is happening. 

I have a RoomMachine and in my Jest tests I make sure to recreate a new interpreter for each test. However it seems as though all of these are somehow linked together (i.e. pointing to the same variable), resulting in failed tests.

I've created a small unit test to demonstrate a weird bit of behaviour:
```
  it('testing', async () => {
    const gen = mockPlayerGen();

    const service1 = interpret(RoomMachine).start();
    const service2 = interpret(RoomMachine).start();

    service1.send('PLAYER_JOIN', { player: gen.next().value });

    console.log(service1.state.context);
    console.log(service2.state.context);

    expect(false).toEqual(true);
  });
```

The output of this when run in jest is:
```
  console.log src/games/machines/room-machine.spec.ts:119
    {
      players: [],
      spectators: [ { socketId: '1', username: 'a', displayUsername: 'A' } ]
    }

  console.log src/games/machines/room-machine.spec.ts:120
    {
      players: [],
      spectators: [ { socketId: '1', username: 'a', displayUsername: 'A' } ]
    }
```

I'm expecting only service1 to have the spectator data... For some reason, sending an event to `service1` is affecting `service2`. I was unable to reproduce this bug in a clean node app. I ran the test file with `npm run test:watch`, which just launches jest in watch mode. 

Here's the test file: [LINK](https://github.com/vck3000/ProAvalon/blob/c18630ecd9ee77dcf60d9fc5a7aba654ac7bcbf8/packages/backend/src/games/machines/room-machine.spec.ts#L111)

Here's the file containing the machine: [LINK](https://github.com/vck3000/ProAvalon/blob/newPA_game_ecs/packages/backend/src/games/machines/room-machine.ts)

I'm using the NestJS framework with Jest, but that shouldn't be a problem I hope... 

Do you guys have any suggestions on what I could try to solve this?

**Comment by davidkpiano** [comment, marked as answer]
⚠️ Do not mutate `context`!

Seems like you're doing it here:

```js
context.spectators.push(event.player);
```

As well as a few other places, probably.

**Reply by vck3000** [reply]
Wow, thank you so much! This fixed my problems. I'll be careful in the future and will make sure to re-read each warning you have in the docs.

**Reply by davidkpiano** [reply]
I should make the warning more clear, to be fair. I'll make a note to do that.
